src/main/scala/org/constellation/domain/redownload/RedownloadService.scala

```diff
@@ -1,38 +1,130 @@
+import org.constellation.util.{HealthChecker, MajorityStateChooser}
 
-class RedownloadService[F[_]]()(implicit F: Concurrent[F]) {
+import scala.concurrent.duration._
+
+class RedownloadService[F[_]](dao: DAO)(implicit F: Concurrent[F], C: ContextShift[F]) {
+  val logger = Slf4jLogger.getLogger[F]
 
   // TODO: Consider Height/Hash type classes
-  private[redownload] val ownSnapshots: Ref[F, Map[Long, String]] = Ref.unsafe(Map.empty)
-  private[redownload] val peersProposals: Ref[F, Map[Long, (Id, String)]] = Ref.unsafe(Map.empty)
+  private[redownload] val ownSnapshots: Ref[F, Map[Long, RecentSnapshot]] = Ref.unsafe(Map.empty)
+  private[redownload] val peersProposals
+    : Ref[F, Map[Long, Seq[NodeSnapshots]]] = Ref.unsafe(Map.empty) //todo join/leave?
 
-  def persistOwnSnapshot(height: Long, hash: String): F[Unit] =
+  def persistOwnSnapshot(height: Long, recentSnapshot: RecentSnapshot): F[Unit] =
     ownSnapshots.modify { m =>
-      val updated = if (m.get(height).nonEmpty) m else m.updated(height, hash)
+      val updated = if (m.get(height).nonEmpty) m else m.updated(height, recentSnapshot)
       (updated, ())
     }
 
-  def getOwnSnapshots(): F[Map[Long, String]] = ownSnapshots.get
+  def getOwnSnapshots(): F[Map[Long, RecentSnapshot]] = ownSnapshots.get
 
-  def getOwnSnapshot(height: Long): F[Option[String]] =
+  def getOwnSnapshot(height: Long): F[Option[RecentSnapshot]] =
     ownSnapshots.get.map(_.get(height))
 
-  // TODO: Use cluster peers to fetch "/snapshot/own" and merge with peersProposals Map
-  def fetchPeersProposals(): F[Unit] = ???
+  def getProposals() =
+    for {
+      peers <- LiftIO[F].liftIO(dao.readyPeers)
+      chunkedPeerProposalMaps <- peers.toList.traverse {
+        case (id, peerData) =>
+          val resp = peerData.client.getNonBlockingFLogged[F, Array[Array[Byte]]](
+            "snapshot/own",
+            timeout = 45.seconds, //todo change these?
+            tag = "snapshot/obj/snapshot"
+          )(C)
+          resp.map((id, _))
+      }
+    } yield chunkedPeerProposalMaps
+
+  def updatePeerProps(fetchedProposals: Seq[(Id, RecentSnapshot)]) =
+    peersProposals.modify { m =>
+      val updatedProposals = fetchedProposals.foldLeft(m) {
+        case (prevPeerProps, (id, recentSnap)) =>
+          val proposalsAtHeight: Seq[(Id, Seq[RecentSnapshot])] = prevPeerProps.getOrElse(recentSnap.height, Seq())
+          val hasMadeProposal = proposalsAtHeight.exists(_._1 == id) //todo Observation of new proposal
+          if (!hasMadeProposal) prevPeerProps.updated(recentSnap.height, proposalsAtHeight :+ (id, Seq(recentSnap)))
+          else prevPeerProps
+      }
+      (updatedProposals, updatedProposals)
+    }
 
-  // TODO: Check for the majority snapshot and store under a variable
-  def recalculateMajoritySnapshot(): F[Unit] = ???
+  def fetchAndSetPeersProposals() = {
+    val fetchedProposals = getProposals().map { proposals =>
```

Comment:
Why it's assigned to a val if this variable is used once only, one line below?

Reply:
We can make it a one liner but I'm of the opinion that it separates the functionality of the method